_posts/2018-07-20-exploring-state-management-in-react-container-components.md

```diff
@@ -0,0 +1,125 @@
+
+```jsx
+class Cart extends React.Component {
+  state = {
+    newItem: ""
+  };
+
+  handleAddItemClick = () => {
+    this.props.addItem(this.state.newItem);
+
+    this.setState({ newItem: "" });
+  };
+
+  render() {
+    return (
+      <ul>
+        <h2>Cart</h2>
+        <input
+          type="text"
+          onChange={({target:{value: newItem}}) => 
+            this.setState({newItem})}
+          value={this.state.newItem}
+        />
+        <button onClick={this.handleAddItemClick}>
+          Add item
+        </button>
+
+        {this.props.items.map((item, idx) => (
+          <li key={idx}>
+            <button onClick={this.props.removeItem(item)}>
+              Remove
+            </button>
+            {item}, &ensp;
+            {this.props.discounts[idx]}% off!
+          </li>
+        ))}
+      </ul>
+    );
+  }
+}
+
+class CartContainer extends React.Component {
+  state = {
+    items: [],
+    discounts: []
+  };
+
+  componentDidUpdate(prevProps, prevState) {
+    if (prevState.items.length !== this.state.items.length) {
+      const itemsQuery = this.state.items.join(",");
+
+      ajax(`https://discountdb.com/?items=${itemsQuery}`)
+        .then(discounts => this.setState({ discounts }));
+    }
+  }
+
+  addItem = item => {
+    this.setState(prevState => ({
+      items: [...prevState.items, item]
+    }));
+  };
+
+  removeItem = item => () => {
+    this.setState(prevState => ({
+      items: prevState.items.filter(
+        (i, idx, prevItems) => idx !== prevItems.indexOf(item)
+      )
+    }));
+  };
+
+  render() {
+    return (
+      <div>
+        <Cart
+          items={this.state.items}
+          discounts={this.state.discounts}
+          addItem={this.addItem}
+          removeItem={this.removeItem}
+        />
+      </div>
+    );
+  }
+}
+```
+&nbsp;
+
+## Benefits
+
+Partitioning your components on the state boundary layer will help reduce complexity by simply having less code to work with at a time, while still staying "inside React".
+
+I really think this pattern starts to pay dividends when you have a lifecycle method, like `componentDidUpdate`, doing a lot of asynchronous work (like making HTTP requests). Given the asynchronous nature, this sort of code tends to be very difficult to test (with both automated unit test and manual testing), so breaking this stateful logic into separate components helps keeps you sane and your code focused. 
+
+It's helpful to remind yourself that when unit testing a React component, you are essentially testing the `render` function. Given the inputs (`props`), what is the output? You've probably noticed tests are painful to write if there is a lot of setup, especially if the setup is required for a feature of the component that you aren't even testing.
+
+Keeping your components small and focused will go a long way for keeping yourself happy and productive!
+
+Play with the code by following the link below!
+
```

Comment:
It feels strange to me not to have a link to Abramov's article on the same topic in here: https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0

Reply:
Good call, I didn't even remember he had written about an article about this (I associate this pattern with you! 😆).